syntax = "proto3";

package heyp.proto;

option go_package = "github.com/uluyol/heyp-agents/go/proto";

message HdrHistogram {
  message Config {
    int64 lowest_discernible_value = 1;
    int64 highest_trackable_value = 2;
    int32 significant_figures = 3;
  }
  message Bucket {
    int64 v = 1; // value
    int64 c = 2; // count
  }
  Config config = 1;
  repeated Bucket buckets = 2;
}

message StatsRecord {
  string label = 1;
  string timestamp = 2; // ISO 8601
  double dur_sec = 3;

  // cumulative stats
  int64 cum_num_bits = 10;
  int64 cum_num_rpcs = 11;

  // current stats
  double mean_bits_per_sec = 21;
  double mean_rpcs_per_sec = 22;
  HdrHistogram latency_ns_hist = 23;

  // convenience percentiles
  int64 latency_ns_p50 = 31;
  int64 latency_ns_p90 = 32;
  int64 latency_ns_p95 = 33;
  int64 latency_ns_p99 = 34;
}
